Sources/XCTest/Public/XCTestCase.swift

```diff
@@ -108,41 +126,185 @@ open class XCTestCase: XCTest {
     /// - Note: Like the designated initializer for Apple XCTest's XCTestCase,
     ///   `-[XCTestCase initWithInvocation:]`, it's rare for anyone outside of
     ///   XCTest itself to call this initializer.
-    public required init(name: String, testClosure: @escaping XCTestCaseClosure) {
+//    @available(*, deprecated, message: "Use init(name:closureType:) instead.")
+    public init(name: String, testClosure: @escaping XCTestCaseClosure) {
         _name = "\(type(of: self)).\(name)"
-        self.testClosure = testClosure
+        self.testClosure = .regular(testClosure)
     }
 
+    public required init(name: String, closureType: ClosureType) {
+        _name = "\(type(of: self)).\(name)"
+        self.testClosure = closureType
+    }
     /// Invoking a test performs its setUp, invocation, and tearDown. In
     /// general this should not be called directly.
     open func invokeTest() {
-        performSetUpSequence()
+        func handleError(_ error: Error) {
+            if error.xct_shouldRecordAsTestFailure {
+                recordFailure(for: error)
+            }
+
+            if error.xct_shouldRecordAsTestSkip {
+                if let skip = error as? XCTSkip {
+                    self.skip = skip
+                } else {
+                    self.skip = XCTSkip(error: error, message: nil, sourceLocation: nil)
+                }
+            }
+        }
 
         do {
             if skip == nil {
-                try testClosure(self)
+                try unwrapAndInvokeTestClosure()
             }
         } catch {
+            handleError(error)
+        }
+
+        if let skip = skip {
+            testRun?.recordSkip(description: skip.summary, sourceLocation: skip.sourceLocation)
+        }
+    }
+
+    private func unwrapAndInvokeTestClosure() throws {
+        let expectation = XCTestExpectation(description: "invokeTest completion")
+
+        @Sendable func handleError(_ error: Error) {
+            
+            if error.xct_shouldSkipTestInvocation && self.skip == nil {
+                if let skip = error as? XCTSkip {
+                    self.skip = skip
+                } else {
+                    self.skip = XCTSkip(error: error, message: nil, sourceLocation: nil)
+                }
+            }
+
             if error.xct_shouldRecordAsTestFailure {
-                recordFailure(for: error)
+                self.recordFailure(for: error)
+            }
+        }
+        
+        switch testClosure {
+        case .regular(let closure):
+            
+            do {
+                try setUpWithError()
+            } catch {
+                handleError(error)
+                return
             }
+            
+            setUp()
+            
+            try closure(self)
+            
+            do {
+                try tearDownWithError()
+            } catch {
+                handleError(error)
+                return
+            }
+            
+            tearDown()
```

Comment:
Need to invoke `tearDown` _before_ `tearDownWithError`: the tear down methods are invoked in the opposite order as the set up methods